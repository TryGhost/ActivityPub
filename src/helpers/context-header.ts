export function getTraceAndSpanId(traceContext: string | undefined) {
    // Get the `X-Cloud-Trace-Context` header from the incoming headers
    if (!traceContext) {
        return { traceId: null, spanId: null };
    }

    // Split by "/" to separate TRACE_ID and SPAN_ID
    const [traceId, spanIdPart] = traceContext.split('/');

    if (traceId && spanIdPart) {
        const spanId = spanIdPart.split(';')[0];
        return { traceId, spanId };
    }

    return { traceId: null, spanId: null };
};
