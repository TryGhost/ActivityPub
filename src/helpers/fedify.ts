import type { Federation } from '@fedify/fedify';

import type { ContextData } from '@/app';

export function createFedifyCtxForHost(
    fedify: Federation<ContextData>,
    host: string,
    ctxData: ContextData,
) {
    let hostUrl: URL;

    try {
        hostUrl = new URL(`https://${host}`);
    } catch (_error) {
        throw new Error(`Invalid host URL: https://${host}`);
    }

    return fedify.createContext(hostUrl, {
        globaldb: ctxData.globaldb,
        logger: ctxData.logger,
    });
}
