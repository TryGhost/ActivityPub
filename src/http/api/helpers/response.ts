function jsonResponse(message: string, status: number) {
    const body = message ? JSON.stringify({ message }) : null;

    return new Response(body, {
        headers: {
            'Content-Type': 'application/json',
        },
        status,
    });
}

export const BadRequest = (message: string) => jsonResponse(message, 400);
export const Forbidden = (message: string) => jsonResponse(message, 403);
export const NotFound = (message: string) => jsonResponse(message, 404);
export const Conflict = (message: string) => jsonResponse(message, 409);
export const InternalServerError = (message: string) =>
    jsonResponse(message, 500);
