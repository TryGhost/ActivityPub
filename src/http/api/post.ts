import type { KnexAccountRepository } from 'account/account.repository.knex';
import { parseURL } from 'core/url';
import type { KnexPostRepository } from 'post/post.repository.knex';
import type { PostService } from 'post/post.service';
import type { AppContext } from '../../app';
/**
 * Create a handler for a request to delete a post
 */
export function createDeletePostHandler(
    accountRepository: KnexAccountRepository,
    postRepository: KnexPostRepository,
    postService: PostService,
) {
    /**
     * Handle a request to delete a post
     */
    return async function handleDeletePost(ctx: AppContext) {
        const logger = ctx.get('logger');

        const id = ctx.req.param('id');

        const idAsUrl = parseURL(id);

        if (!idAsUrl) {
            return new Response(null, {
                status: 400,
            });
        }

        const account = await accountRepository.getBySite(ctx.get('site'));
        const post = await postService.getByApId(idAsUrl);

        if (!post) {
            return new Response(null, {
                status: 404,
            });
        }

        if (post.author.uuid !== account.uuid) {
            return new Response(null, {
                status: 403,
            });
        }

        try {
            post.delete(account);
            await postRepository.save(post);

            return new Response(null, {
                status: 204,
            });
        } catch (err) {
            logger.error('Error deleting post - {error}', {
                error: err,
            });
            return new Response(JSON.stringify(err), { status: 500 });
        }
    };
}
