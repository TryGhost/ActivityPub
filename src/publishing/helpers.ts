import {
    FedifyActivitySender,
    FedifyActorResolver,
    FedifyKvStoreObjectStore,
    FedifyKvStoreOutbox,
    FedifyUriBuilder,
} from '../activitypub';
import { type AppContext, fedify } from '../app';
import { FedifyPublishingService } from './service';
import type { Post } from './types';

/**
 * Publish a post to the Fediverse
 *
 * @param ctx App context instance
 * @param post Post to publish
 */
export async function publishPost(ctx: AppContext, post: Post) {
    const scopedDb = ctx.get('db');
    const globalDb = ctx.get('globaldb');
    const logger = ctx.get('logger');

    const fedifyCtx = fedify.createContext(ctx.req.raw, {
        db: scopedDb,
        globaldb: globalDb,
        logger: ctx.get('logger'),
    });

    const publishingService = new FedifyPublishingService(
        new FedifyActivitySender(fedifyCtx),
        new FedifyActorResolver(fedifyCtx),
        logger,
        new FedifyKvStoreObjectStore(globalDb),
        new FedifyUriBuilder(fedifyCtx),
    );

    const outbox = new FedifyKvStoreOutbox(scopedDb);

    await publishingService.publishPost(post, outbox);
}
