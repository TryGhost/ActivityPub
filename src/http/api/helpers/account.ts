import { type Context, isActor } from '@fedify/fedify';
import type { Account } from 'account/account.entity';
import type { AccountService } from 'account/account.service';
import type { Site } from 'account/types';
import { getAccountHandle } from 'account/utils';
import type { ContextData } from 'app';
import {
    getAttachments,
    getFollowerCount,
    getFollowingCount,
    getHandle,
    isFollowedByDefaultSiteAccount,
} from 'helpers/activitypub/actor';
import { sanitizeHtml } from 'helpers/html';
import { lookupObject } from 'lookup-helpers';
import type { AccountDTO } from '../types';

/**
 * Converts an Account to an AccountDTO
 * @param account - The account to convert
 *
 * @returns Promise resolving to AccountDTO
 */
export async function getAccountDtoFromAccount(
    account: Account,
    defaultAccount: Account,
    accountService: AccountService,
): Promise<AccountDTO> {
    const accountDto: AccountDTO = {
        id: account.id?.toString() || '',
        name: account.name || '',
        handle: getAccountHandle(new URL(account.apId).host, account.username),
        bio: sanitizeHtml(account.bio || ''),
        url: account.url.toString() || '',
        avatarUrl: account.avatarUrl?.toString() || '',
        /**
         * At the moment we don't support banner images for Ghost accounts
         */
        bannerImageUrl: account.bannerImageUrl?.toString() || '',
        /**
         * At the moment we don't support custom fields for Ghost accounts
         */
        customFields: {},
        attachment: [],
        postCount: await accountService.getPostCount(account.id),
        likedCount: await accountService.getLikedCount(account.id),
        followingCount: await accountService.getFollowingAccountsCount(
            account.id,
        ),
        followerCount: await accountService.getFollowerAccountsCount(
            account.id,
        ),
        followedByMe: await accountService.checkIfAccountIsFollowing(
            defaultAccount.id,
            account.id,
        ),
        followsMe: false,
    };
    return accountDto;
}

/**
 * Retrieves an AccountDTO by looking up an actor using their handle
 * @param handle - The handle to look up (e.g., "@username@domain")
 *
 * @returns Promise resolving to AccountDTO
 */
export async function getAccountDTOByHandle(
    handle: string,
    apCtx: Context<ContextData>,
    site: Site,
    accountService: AccountService,
): Promise<AccountDTO> {
    if (!handle) {
        throw new Error('Handle is null');
    }

    // Lookup actor by handle
    const actorObj = await lookupObject(apCtx, handle);

    if (!isActor(actorObj)) {
        throw new Error('Actor not found');
    }

    const actor: any = await actorObj.toJsonLd();

    const [followerCount, followingCount, isFollowingResult, attachments] =
        await Promise.all([
            getFollowerCount(actorObj),
            getFollowingCount(actorObj),
            isFollowedByDefaultSiteAccount(actorObj, site, accountService),
            getAttachments(actorObj, {
                sanitizeValue: (value: string) => sanitizeHtml(value),
            }),
        ]);

    const accountDto: AccountDTO = {
        id: actor.id,
        name: actor.name || '',
        handle: getHandle(actorObj),
        bio: sanitizeHtml(actor.summary),
        url: actor.url || '',
        avatarUrl: actor.icon?.url || '',
        bannerImageUrl: actor.image?.url || '',
        customFields: {},
        postCount: 0,
        likedCount: 0,
        followingCount: followingCount,
        followerCount: followerCount,
        followedByMe: isFollowingResult,
        followsMe: false,
        attachment: attachments,
    };
    return accountDto;
}
