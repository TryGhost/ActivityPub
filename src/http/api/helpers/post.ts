import { isActor } from '@fedify/fedify';
import type { Account } from 'account/account.entity';
import type { Post } from 'post/post.entity';
import type { AccountService } from '../../../account/account.service';
import {
    getAccountHandle,
    mapActorToExternalAccountData,
} from '../../../account/utils';
import type { FedifyRequestContext } from '../../../app';
import {
    ACTIVITY_OBJECT_TYPE_ARTICLE,
    ACTIVITY_OBJECT_TYPE_NOTE,
    ACTIVITY_TYPE_ANNOUNCE,
} from '../../../constants';
import type {
    Activity,
    ActivityObject,
    ActivityObjectAttachment,
} from '../../../helpers/activitypub/activity';
import { lookupActor } from '../../../lookup-helpers';
import { PostType } from '../../../post/post.entity';
import type { AuthorDTO, PostDTO } from '../types';

/**
 * Get the author of a post from an activity: If the activity has attribution,
 * use that as the the author, otherwise use the activity actor as the author
 *
 * @param activity Activity
 * @param accountService Account service instance
 * @param fedifyCtx Fedify request context instance
 */
export async function getPostAuthor(
    activity: Activity,
    accountService: AccountService,
    fedifyCtx: FedifyRequestContext,
) {
    let activityPubId: string;

    if (typeof activity.actor === 'string') {
        activityPubId = activity.actor;
    } else {
        activityPubId = activity.actor.id;
    }

    const object = activity.object as ActivityObject;

    if (object.attributedTo && typeof object.attributedTo === 'string') {
        activityPubId = object.attributedTo;
    }

    if (object.attributedTo && typeof object.attributedTo === 'object') {
        activityPubId = object.attributedTo.id;
    }

    let author = await accountService.getAccountByApId(activityPubId);

    // If we can't find an author, and the activity is an announce, we need to
    // look up the actor and create a new account, as we may not have created
    // the account yet - currently, accounts only get created when a follow
    // occurs, but the user may not be following the original author of the
    // announced object. This won't be needed when we have the posts table as
    // this enforces that a post belongs to an account (so the account has to
    // exist prior to insertion into the posts table)
    if (!author && activity.type === ACTIVITY_TYPE_ANNOUNCE) {
        const actor = await lookupActor(fedifyCtx, activityPubId);

        if (isActor(actor)) {
            const externalAccountData =
                await mapActorToExternalAccountData(actor);

            author =
                await accountService.createExternalAccount(externalAccountData);
        }
    }

    return author;
}

/**
 * Get the author of a post from an activity without attribution
 *
 * @param activity Activity
 * @param accountService Account service instance
 */
export async function getPostAuthorWithoutAttribution(
    activity: Activity,
    accountService: AccountService,
) {
    let activityPubId: string;

    if (typeof activity.actor === 'string') {
        activityPubId = activity.actor;
    } else {
        activityPubId = activity.actor.id;
    }

    return accountService.getAccountByApId(activityPubId);
}

/**
 * Get the excerpt of a post from an activity
 *
 * @param activity Activity
 */
export function getPostExcerpt(activity: Activity) {
    const object = activity.object as ActivityObject;

    if (object.type === ACTIVITY_OBJECT_TYPE_NOTE) {
        return '';
    }

    if (object.preview) {
        return object.preview.content ?? '';
    }

    return object.content
        .replace(/<[^>]*>/g, ' ') // Swap out HTML tags with a space
        .replace(/\s+/g, ' ') // Normalize all whitespace (newlines, multiple spaces) to single space
        .trim()
        .slice(0, 400);
}

/**
 * Get the feature image URL for a post from an activity
 *
 * @param activity Activity
 */
export function getPostFeatureImageUrl(activity: Activity): string | null {
    const object = activity.object as ActivityObject;

    if (typeof object.image === 'string') {
        return object.image;
    }

    if (typeof object.image === 'object' && object.image.url) {
        return object.image.url;
    }

    return null;
}

/**
 * Compute the reading time (in minutes) for a post's content
 *
 * @param content Post content
 */
export function getPostContentReadingTimeMinutes(content: string) {
    const WORDS_PER_MINUTE = 275;

    if (!content) return 0;

    const plainTextContent = content
        .replace(/<[^>]*>/g, '') // Remove HTML tags
        .replace(/&[^;]+;/g, ' ') // Remove HTML entities
        .trim();

    const wordCount = plainTextContent
        .split(/\s+/)
        .filter((word) => word.length > 0).length;

    return Math.ceil(wordCount / WORDS_PER_MINUTE);
}

/**
 * Get the attachments for a post from an activity
 *
 * @param activity Activity
 */
export function getPostAttachments(
    activity: Activity,
): ActivityObjectAttachment[] {
    const object = activity.object as ActivityObject;

    return (
        Array.isArray(object.attachment)
            ? object.attachment
            : [object.attachment].filter(
                  (attachment) => attachment !== undefined,
              )
    ).map((attachment) => {
        return {
            type: attachment.type,
            mediaType: attachment.mediaType,
            name: attachment.name,
            url: attachment.url,
        };
    });
}

/**
 * Map an Activity to a Post
 *
 * @param activity Activity
 * @param accountService Account service instance
 * @param fedifyCtx Fedify request context instance
 */
export async function mapActivityToPost(
    activity: Activity,
    accountService: AccountService,
    fedifyCtx: FedifyRequestContext,
): Promise<PostDTO | null> {
    const object = activity.object as ActivityObject;

    // At the moment it is possible that a post can be published without any
    // content, so we need to handle this case by using an empty string
    const postContent = object.content || '';

    const author = await getPostAuthor(activity, accountService, fedifyCtx);

    // If we can't find an author, we can't map the activity to a post, so we
    // return early
    if (!author) {
        return null;
    }

    const post: PostDTO = {
        id: object.id,
        type:
            object.type === ACTIVITY_OBJECT_TYPE_ARTICLE
                ? PostType.Article
                : PostType.Note,
        title: object.name || '', // A note doesn't have a title
        excerpt: getPostExcerpt(activity),
        content: postContent,
        url: object.url,
        featureImageUrl: getPostFeatureImageUrl(activity),
        // When the activity is an announce, we want to use the published date of
        // the announce rather than the published date of the object
        publishedAt:
            activity.type === ACTIVITY_TYPE_ANNOUNCE
                ? activity.published
                : object.published,
        // `buildActivity` adds a `liked` property to the object if it
        // has been liked by the current user
        likeCount: object.liked ? 1 : 0,
        likedByMe: Boolean(object.liked),
        // `buildActivity` adds a `replyCount` property to the object
        replyCount: object.replyCount || 0,
        readingTimeMinutes: getPostContentReadingTimeMinutes(postContent),
        attachments: getPostAttachments(activity),
        author: {
            id: author.id.toString(),
            handle: getAccountHandle(
                new URL(author.ap_id).host,
                author.username,
            ),
            avatarUrl: author.avatar_url ?? '',
            name: author.name ?? '',
            url: author.url ?? '',
        },
        repostCount: object.repostCount ?? 0,
        repostedByMe: Boolean(object.reposted),
        repostedBy: null,
    };

    if (activity.type === ACTIVITY_TYPE_ANNOUNCE) {
        const repostedBy = await getPostAuthorWithoutAttribution(
            activity,
            accountService,
        );

        if (repostedBy) {
            post.repostedBy = {
                id: repostedBy.id.toString(),
                handle: getAccountHandle(
                    new URL(repostedBy.ap_id).host,
                    repostedBy.username,
                ),
                avatarUrl: repostedBy.avatar_url ?? '',
                name: repostedBy.name ?? '',
                url: repostedBy.url ?? '',
            };
            post.repostCount = 1;
        }
    }

    return post;
}

function accountToAuthorDTO(account: Account): AuthorDTO {
    return {
        id: account.apId.href,
        name: account.name || '',
        handle: account.username,
        avatarUrl: account.avatarUrl?.href || '',
        url: account.url.href,
    };
}

export function postToDTO(
    post: Post,
    meta: {
        likedByMe: boolean;
        repostedByMe: boolean;
        repostedBy: Account | null;
    } = {
        likedByMe: false,
        repostedByMe: false,
        repostedBy: null,
    },
): PostDTO {
    return {
        id: post.apId.href,
        type: post.type,
        title: post.title ?? '',
        excerpt: post.excerpt ?? '',
        content: post.content ?? '',
        url: post.url.href,
        featureImageUrl: post.imageUrl?.href ?? null,
        publishedAt: post.publishedAt,
        likeCount: post.likeCount,
        likedByMe: meta.likedByMe,
        replyCount: post.replyCount,
        readingTimeMinutes: post.readingTimeMinutes,
        attachments: post.attachments.map((attachment) => {
            return {
                name: attachment.name ?? '',
                type: attachment.type ?? '',
                mediaType: attachment.mediaType ?? '',
                url: attachment.url.href,
            };
        }),
        author: accountToAuthorDTO(post.author),
        repostCount: post.repostCount,
        repostedByMe: meta.repostedByMe,
        repostedBy: meta.repostedBy
            ? accountToAuthorDTO(meta.repostedBy)
            : null,
    };
}
