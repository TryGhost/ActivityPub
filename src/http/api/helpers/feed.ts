import type { Account } from '@/account/account.entity';
import { getAccountHandle } from '@/account/utils';
import { parseURL } from '@/core/url';
import type { GetFeedDataResultRow } from '@/feed/feed.service';

export function feedResultToPostDTO(
    results: GetFeedDataResultRow[],
    myAccount: Account,
) {
    return results.map((result) => {
        return {
            id: result.post_ap_id,
            type: result.post_type,
            title: result.post_title ?? '',
            excerpt: result.post_excerpt ?? '',
            summary: result.post_summary ?? null,
            content: result.post_content ?? '',
            url: result.post_url,
            featureImageUrl: result.post_image_url ?? null,
            publishedAt: result.post_published_at,
            likeCount: result.post_like_count,
            likedByMe: result.post_liked_by_user === 1,
            replyCount: result.post_reply_count,
            readingTimeMinutes: result.post_reading_time_minutes,
            attachments: result.post_attachments
                ? result.post_attachments.map((attachment) => ({
                      type: attachment.type ?? '',
                      mediaType: attachment.mediaType ?? '',
                      name: attachment.name ?? '',
                      url: attachment.url,
                  }))
                : [],
            author: {
                id: result.author_id.toString(),
                handle: getAccountHandle(
                    parseURL(result.author_url)?.host ?? '',
                    result.author_username,
                ),
                name: result.author_name ?? '',
                url: result.author_url ?? '',
                avatarUrl: result.author_avatar_url ?? '',
                followedByMe: result.author_followed_by_user === 1,
            },
            authoredByMe: result.author_id === myAccount.id,
            repostCount: result.post_repost_count,
            repostedByMe: result.post_reposted_by_user === 1,
            repostedBy: result.reposter_id
                ? {
                      id: result.reposter_id.toString(),
                      handle: getAccountHandle(
                          parseURL(result.reposter_url)?.host ?? '',
                          result.reposter_username,
                      ),
                      name: result.reposter_name ?? '',
                      url: result.reposter_url ?? '',
                      avatarUrl: result.reposter_avatar_url ?? '',
                      followedByMe: result.reposter_followed_by_user === 1,
                  }
                : null,
        };
    });
}
