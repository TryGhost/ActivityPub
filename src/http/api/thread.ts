import type { AccountService } from 'account/account.service';
import type { AppContext } from '../../app';
import type { KnexPostRepository } from '../../post/post.repository.knex';
import { postToDTO } from './helpers/post';

/**
 * Create a handler for a request for a thread
 *
 * @param postRepository Post repository instance
 * @param accountService Account service instance
 */
export function createGetThreadHandler(
    postRepository: KnexPostRepository,
    accountService: AccountService,
) {
    /**
     * Handle a request for a thread
     *
     * @param ctx App context instance
     */
    return async function handleGetThread(ctx: AppContext) {
        const postApId = decodeURIComponent(ctx.req.param('post_ap_id'));

        const account = await accountService.getDefaultAccountForSite(
            ctx.get('site'),
        );

        const posts = (
            await postRepository.getThreadByApId(postApId, account.id)
        ).map(({ post, likedByAccount, repostedByAccount }) => {
            return postToDTO(post, {
                likedByMe: likedByAccount,
                repostedByMe: repostedByAccount,
                repostedBy: null,
            });
        });

        return new Response(
            JSON.stringify({
                posts,
            }),
            {
                status: 200,
            },
        );
    };
}
