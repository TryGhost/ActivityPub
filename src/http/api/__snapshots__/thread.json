{
  "posts": [
    {
      "attachments": [],
      "author": {
        "avatarUrl": "https://example.com/avatars/foobar.png",
        "handle": "@foobar@example.com",
        "id": "https://example.com/users/456",
        "name": "Foo Bar",
        "url": "https://example.com/users/456",
      },
      "content": "Test Post 123 Content",
      "excerpt": "Test Post 123 Excerpt",
      "featureImageUrl": "https://example.com/images/post-123.jpg",
      "id": "https://example.com/.ghost/activitypub/article/259e92cb-5ac2-4d62-910f-ddea29b2cf55",
      "likeCount": 0,
      "likedByMe": false,
      "publishedAt": "2025-02-27T15:40:00.000Z",
      "readingTimeMinutes": 1,
      "replyCount": 2,
      "repostCount": 0,
      "repostedBy": null,
      "repostedByMe": false,
      "title": "Test Post 123",
      "type": 1,
      "url": "https://example.com/posts/123",
    },
    {
      "attachments": [],
      "author": {
        "avatarUrl": "https://example.com/avatars/foobar.png",
        "handle": "@foobar@example.com",
        "id": "https://example.com/users/456",
        "name": "Foo Bar",
        "url": "https://example.com/users/456",
      },
      "content": "Test Post 456 Content",
      "excerpt": "Test Post 456 Excerpt",
      "featureImageUrl": "https://example.com/images/post-456.jpg",
      "id": "https://example.com/.ghost/activitypub/article/9ac8f7cd-77e8-4abe-b075-e008bd24f3c5",
      "likeCount": 0,
      "likedByMe": true,
      "publishedAt": "2025-02-27T15:40:00.000Z",
      "readingTimeMinutes": 1,
      "replyCount": 0,
      "repostCount": 0,
      "repostedBy": null,
      "repostedByMe": false,
      "title": "Test Post 456 (reply to Test Post 123)",
      "type": 1,
      "url": "https://example.com/posts/456",
    },
    {
      "attachments": [],
      "author": {
        "avatarUrl": "https://example.com/avatars/foobar.png",
        "handle": "@foobar@example.com",
        "id": "https://example.com/users/456",
        "name": "Foo Bar",
        "url": "https://example.com/users/456",
      },
      "content": "Test Post 789 Content",
      "excerpt": "Test Post 789 Excerpt",
      "featureImageUrl": "https://example.com/images/post-789.jpg",
      "id": "https://example.com/.ghost/activitypub/article/e2cfad55-6275-46e8-934f-7fee70fa79f5",
      "likeCount": 0,
      "likedByMe": false,
      "publishedAt": "2025-02-27T15:40:00.000Z",
      "readingTimeMinutes": 1,
      "replyCount": 0,
      "repostCount": 0,
      "repostedBy": null,
      "repostedByMe": true,
      "title": "Test Post 789 (reply to Test Post 123)",
      "type": 1,
      "url": "https://example.com/posts/789",
    },
  ],
}